pep-0690.rst

```diff
@@ -312,57 +325,54 @@ import object. When ``modb.foo`` is later referenced, it will also try to
+
+``PyDict_Next`` will attempt to resolve all lazy import objects the first time
+position ``0`` is accessed, and those imports could fail with exceptions. Since
+``PyDict_Next`` cannot set an exception, ``PyDict_Next`` will return ``0``
+immediately in this case, and any exception will be printed to stderr as an
+unraisable exception.
+
+For this reason, this PEP introduces ``PyDict_NextWithError``, that works in the
+same way as ``PyDict_Next``, but which can set an error when returning ``0`` and
+this should be checked via ``PyErr_Occurred()`` after the call.
 
 The eagerness of imports within ``try`` / ``except`` / ``with`` blocks or within
 class or function bodies is handled in the compiler via a new
 ``EAGER_IMPORT_NAME`` opcode that always imports eagerly. Top-level imports use
 ``IMPORT_NAME``, which may be lazy or eager depending on ``-L`` and/or
 ``importlib.set_lazy_imports()``.
 
-
-Exceptions
-----------
-
-Exceptions that occur during a lazy import bubble up and erase the
-partially-constructed module(s) from ``sys.modules``, just as exceptions during
-normal import do.
-
-Since errors raised during a lazy import will occur later (wherever the imported
-name is first referenced) than they would if the import were eager, it is
-possible that they could be accidentally caught by exception handlers that
-didn't expect the import to be running within their ``try`` block, leading to
-confusion. To reduce the potential for this confusion, exceptions raised in the
-course of executing a lazy import will be replaced by a ``LazyImportError``
-exception (a subclass of ``ImportError``), with ``__cause__`` set to the
-original exception.
-
-The ``LazyImportError`` will have source location metadata attached pointing the
-user to the original import statement, to ease debuggability of errors from lazy
-imports. (It won't have a full traceback to the original import location; this
-is too expensive to preserve for all lazy imports, and it's not clear that it
-provides significant value over simply knowing the location of the import
-statement.)
-
-Only ``Exception`` are replaced in this way, not ``BaseException``.
-``BaseException`` are for "system-exiting" exceptions like ``KeyboardInterrupt``
-or ``SystemExit``; these are normally not caught, and if they are caught, it is
-less likely to be specific to a certain bit of code that was expected to raise
-them, and more likely that the goal is to catch them whatever their origin.
+The current status of lazy imports at any given place is determined by a
+combination of the passed ``-L`` option flag; an interpreter-wide flag set by
+``importlib.set_lazy_imports()`` and the container object passed in its
+``excluding`` keyword argument; and a flag set by
+``importlib.enable_lazy_imports_in_module()`` in nearest running module frame.
+All these together are used to cache the current status of lazy imports in the
+currently running frame. This cache is globally busted whenever any of these API
```

Comment:
Does it seem realistic that any code can actually perform these checks in a reasonably easy or readable way?  Do you think any code would need to do these calculations?  If the answers to these is "no" and "yes", would it make sense to add a convenience function that would encapsulate all of these checks in one call?

Reply:
Oh! this is only how it is implemented, for transparency of what's going on. The function to use for checking if lazy imports is enabled at any given point is `importlib. is_lazy_imports_enabled()`, that already does all this checking and handles the cache for it :)

Reply:
I added that there to make it clearer.